resources/type-system/inference.md

```diff
@@ -1210,6 +1247,674 @@ of steps:
+
+_For the example above, C and D are condensed into a single node, so the graph
+now looks like this:_
+
+&emsp;{B} &lArr; {A} &rArr; {C, D}
+
+Finally, the nodes are grouped into stages as follows. Stage zero consists of
+all nodes that have no outgoing edges. Then, those nodes are removed from the
+graph, along with all their incoming edges. For each of the following stages,
+the same procedure is followed: from the graph produced by stage _k_, let stage
+_k+1_ be the set of nodes that have no outgoing edges, then delete those nodes
+and their incoming edges to produce a newly reduced graph. This is repeated
+until the graph is empty.
+
+_In this example, that means that there will be two stages. The first stage will
+consist of arguments B, C, and D, and the second stage will consist of argument
+A._
+
+_The intuitive justification for this algorithm is that by condensing the
+dependency graph into strongly connected components, we ensure that, in the
+absence of dependency cycles, dependency arcs always go from earlier stages to
+later stages; in other words we obtain each bit of information before it is
+needed, if possible. In the event that it's not possible due to a dependency
+cycle, we group all arguments in the dependency cycle into the same stage, which
+reproduces the behavior of the Dart language before horizontal inference was
+introduced._
+
+_(Note that the property mentioned earlier, that non-function literals are
+always placed in the first stage, is guaranteed by the fact that dependency
+analysis only draws an edge from A to B if A is a function literal.)_
+
+## Method invocation inference
+
+The type inference rules for multiple kinds of expressions make use of an
+operation known as _method invocation inference_. _Method invocation inference_
+is a type inference step that is applied to a target elaborated expression
+`m_0`, a method name identifier or operator name `id`, a sequence of expressions
+`{e_1, e_2, ...}`, a corresponding sequence (of the same length) of optional
+name identifiers `{n_1, n_2, ...}`, a sequence of zero or more type arguments
+`{T_1, T_2, ...}`, and a type schema `K`. As with [argument part
+inference](#Argument-part-inference), the symbol `∅` denotes a missing name.
+
+The output of method invocation inference is an elaborated expression `m`, with
+static type `T`, where `m` and `T` are determined as follows:
+
+- Let `T_0` be the static type of `m_0`.
+
+- If `T_0` is `void`, there is a compile-time error.
+
+- Let `U_0` be the [bound resolution](#Bound-resolution) of `T_0`.
+
+- If `U_0` is `dynamic` or `Never`, or `U_0` is `Function` and `id` is `call`,
+  then:
+
+  - Invoke [argument part inference](#Argument-part-inference) on `<T_1, T_2,
+    ...>(n_1: e_1, n_2: e_2, ...)`, using a target function type of `∅` and a
+    type schema `K`. Denote the resulting elaborated expressions by `{m_1, m_2,
```

Comment:
Saying "a type schema `K`" is a little confusing because it has two natural interpretations: "some type schema K (I'm not saying what)" and "the type schema K that I mentioned above".  I would either say "the type schema K" or "K as the type schema" to make it clear that `K` is not being bound here but rather is a reference to the K introduced above.

Reply:
Same comment for each following bullet.

Reply:
Good point. I went with “using `∅` as the target function type and `K` as the context”.